### ISSUE TITLE
Security and the general system architecture

### ISSUE THREAD

**Issue by ndim** [lack/soundcraft-utils#8]
As far as I can tell, `soundcraft-utils` operates in either of the
following three architecture variants:

```
    non-root                    root
    user process                system user process         kernel

(1) soundcraft_ctl              soundcraft_dbus_service     /dev/bus/usb/003/003
(2) soundcraft_ctl --no-dbus                                /dev/bus/usb/003/003
(3) soundcraft_gui              soundcraft_dbus_service     /dev/bus/usb/003/003
```

A process running as `root` is always a bad idea from a security
POV.

Why does `soundcraft-utils` need `root` priviledges? User access to
USB devices can be set up with udev/polkit/etc rules, even in a
non-user specific way, e.g. with the udev action `TAG+="uaccess"` for
the user currently logged in locally.

As the `soundcraft_dbus_service` process, running the soundcraft.dbus
module code, also contains code to install file system wide, this
makes the process an interesting target for attacks, e.g. for local
priviledge escalation. As I learned again in
https://github.com/johnath/beep/issues/11 it is very difficult to
consider each and every attack surface for a process running as root,
so I cannot tell right off the top of my head what that surfaces might
be. However, Python contains a LOT of code which might have issues,
and libusb being C code might also become a problem some time.

I personally would like to make both the `soundcraft_ctl` CLI and
`soundcraft_gui` GUI compatible with direct USB access, and either get
rid of the dbus service running as root, or at least make it optional
for the `soundcraft_ctl` and `soundcraft_gui` programs to run through
the soundcraft_dbus_service.

Then the `soundcraft-utils` sources could ship a few example
permission setup rules for udev and related mechanisms, and maybe also
install at least the generic ones with `python setup.py install`.

**Comment by lack** [comment]
I totally agree that the security model leaves much to be desired, and needs improvement.

I have a dbus service running as root because it was the quickest way forward to get something working without worrying too much about per-distro quirks and packaging/installation issues.  I do want to eventually lock this down more, but haven't got around to it yet.  It's a big hammer, but it got the thing out the door to where other people could use it so I could improve it based on real user input, not just my own whims.  (And while we're on the subject, thank you so much for your time and dedication to these issues.  I really do appreciate this feedback!)

I understand your concern about the attack surface, but this is something I did some thinking about when designing it.  The DBUS interface itself is pretty limited.  All but one property is read-only, the only exception being the `routingSource` property that's actually used to change the USB input mapping.  This is parsed pretty carefully into an integer that's used when it's passed down to the USB layer, and never passed through any sort of 'eval' mechanism, so I don't think there's much of a chance for exploit there.

The installation procedure itself is an interesting potential avenue for exploit I hadn't thought a lot about.  I think I list all the files that are installed as part of the '--setup' step so that a concerned administrator could inspect what changes were made to the system, and I suppose a very-concerned administrator could and should inspect the code to find out what '--setup' does before running it as root.  That said, I'm open to any improvements you would suggest here.

I would also not object to adding a '--no-dbus' mode to the GUI as well, if that's something you would like to add.  It would be pretty easy to do by reusing the 'autodetect' code that's in the soundcraft/cli.py in support of the same option.  I think the other thing needed to make this work correctly would be to change the location of the state file in soundcraft/notepad.py to something in $HOME when it's run in no-dbus mode.

That being said, I don't think I want to get rid of the dbus service, though.  Here's what it gives us that I don't want to lose:
- A single source of "truth" about the state of the hardware.  Due to what I think is a flaw in the hardware itself, there doesn't seem to be a way to query the active state of the USB routing, so I need to keep the state locally somewhere.
- A unified access to that hardware which properly handles concurrent access and signalling to multiple consumers.  For example, if I'm running the GUI and then I use the CLI to change the state, DBUS notifies the GUI and it updates things automagically.
- It allows easier external integration.  For example, maybe someone will write a plugin for their favorite DAW that uses the dbus interface.

I think there are two ways I'd like to see the hardening of this dbus service progress, which are not mutually-exclusive:
- Let's not run the service as root, but as a non-privileged user that has access to the appropriate USB devices
- Let's lock down the DBUS interface with better more granular permissions.

**Comment by ndim** [comment]
[A few quick comment from on the road, typed on a small screen without much in the way of overview, so this may be a bit unordered and confused.]

I can see now that the dbus service gives a few things which cannot be achieved otherwise, namely serializing accesses from multiple dbus client processes to the same piece of hardware while keeping those processes' idea of the hardware state and the hardware state in sync.

[Perhaps Soundcraft's OSX utility does something different from the Windows utility, so there might be a state query command to be found there.]

Moving the dbus service to a soundcraft-utils specific non-root user could be a good way to package soundcraft-utils, while the default install from source could very well keep using a root dbus service for simpler cross-distro install and setup.

However, I would like to remove the installation and uninstallation code from the dbus service. Generating the files from templates could be hooked into `setup.py`. Actually hooking the dbus and xdg files into dbus and xdg could then be done by a special soundcraft_setup utility or a special setup.py command `install-dbus` and `install-xdg` which a standard `setup.py install` would then not imply.

At least for xdg `.desktop` files there is a well defined location for system wide installation somewhere in `/usr/share`, and for per user installation somewhere in `$HOME/`, so those two different installation types could be supported relatively easily.

If the dbus service is to run as any special user (whether that is `root` or `soundcraft-utils`), then a standard `pip install --user` will not be able to help there. However, is it still an actual `--user` install if I need to set up dbus services in `/usr/`?

[If the soundcraft-utils is installed with `--user`, and the dbus service runs as root, ... isn't it possible for **any** process running as the non-root user to change the `soundcraft/**/*.py` code files which dbus will then happily run as root when starting the dbus service? I will need to check that, but it sounds both possible and undesirable outside of a specifically development settings.]

As for adding `--no-dbus`, I will take a shot at the code for that.

**Comment by ndim** [comment]
OK, after some discussion on the `#dbus` IRC channel with Simon McV, some re-reading of what we have discussed here, and some reflection on my part, I would recommend that `soundcraft-utils` move the dbus service from the system bus to the session bus.

The current state of `soundcraft-utils` is that the dbus service is run on the system bus as root. This creates two security problems:

  * For system-wide as-root installations to '/usr/local' or '/usr', there might be problems with `soundcraft-utils` dbus service parsing of parameters, or in use of APIs, or whatever. None of these problems would be obvious, and could probably fixed out to a satifactory level. This is maybe a bit hypothetical and speculating on future events, but the next one is very real and present.
  * For non-root installations such as `--user` or `venv`, the code actually run by the dbus service as root is writable by the non-root user, and thus a direct path to local priviledge escalation: User stops the dbus service, changes its code to do whatever, restarts the dbus service, which then gladly runs whatever as root. This is not much of a problem when a user-admin-developer is actively working on the code, but if a mainly-user does it once, is glad that it works, and then forgets about it (as "it is working"), it definitively becomes a problem.

## Stage one

I propose moving the dbus service from the system bus to the service bus.

This means on the feature side, that the three main properties you have listed above are still provided

  * single copy of the hardware state for the user's session
  * unified access to the hardware for processes inside the user's session
  * allows other processes like DAWs inside the user's session access to the hardware state

and on the fixes side

  * no avenue for priviledge escalation in any of the two above scenarios

The possible downside is that the user/admin installing `soundcraft-utils` will now need to add `udev` rules to set up the permissions to the USB device, and that might be a bit non-standard for non-standard distributions. However, the udev rules should not be all *that* system specific these days, thanks to the cross-distribution efforts for desktop integration. The `TAG+="uaccess"` rules should work for most distributions, and the single step of dropping a file to `/etc/udev/rules.d/` will work for all future installations of `soundcraft-utils`, regardless of for which user on the system they happen or whether they happen system wide.

If you want to, you can still write a `soundcraft-setup` utility which can try figuring out the best way to figure out the proper udev rules for the system.

The installation of `$HOME/.local/share/dbus-1/services/*.service` and `$HOME/.local/share/applications/*.desktop` and `$HOME/.local/share/icons/hicolor/**/*.png` would then be hooked into `setup.py install`, so that it works for every type of installation. Note that those files need to be installed outside of any `venv`.

For systems with only one user session running anything related to `soundcraft-utils`, this should work very well even when switching from one logged in user to another.

## Stage 2

If someone wants different DAW sessions running for different users with different audio routing, then the `soundcraft-utils` session dbus service could mimic what the `pulseaudio` session dbus service does regarding releasing and re-taking devices. Then the session dbus service could change the hardware state every time it re-takes a device, and now every session has its own view of the hardware state reflected on the hardware whenever that session is active.

However, there is a conflicting use case.

If someone wants processes from multiple user sessions to access the hardware at the same time and have a common hardware state amongst those different sessions, a system dbus service would need to come into play again. The system dbus service can provide a common view onto the same hardware, and the per-session state keeping would need to be disabled. The session dbus service could be changed to talk to the system dbus service, which then would only run on system-wide installations. The hardware access for the active session user would need to be revoked, and hardware access granted only to the system wide dbus service.

No easy solution of the conflict between those two use cases.

## Stage three

If the system dbus service becomes too cumbersome or does not cover some special cases, its functionality could be moved inside the kernel driver. The kernel's purpose is hardware arbitration, so that would not be very far fetched. However, user space drivers are often preferred these days, and that is what a system dbus service would provide.

## Stage four

Use cases with Notepad mixers being unplugged and replugged to different computers and the software on those different computers needing to use the actual state from the hardware cannot be served with any PC side solution which cannot query the hardware state. Someone will have to change the Notepad firmware to add a USB command to query the hardware state.

## Stage five

When someone gets a soldering iron and a drill into their hands, they could also add one LED per input source to the mixer front panel to display the state, and at least one pushbutton to select the source. Then the whole USB and software controlled setting of the audio routing can be removed.

## Conclusion

Can we just implement stage one and hope that will carry us along for the next twenty years?

**Comment by lack** [comment]
I'm not sure I agree about moving the dbus service to the session bus - This dbus service represents the real hardware which is a singleton.  There's no per-user hardware.  It's globally visible to the running system so the system bus fits that reality better.

So let's talk more about your 2 security concerns, which I think are very important:

1.  _When running on the system bus as root, code errors could lead to privilege escalations._  True.  I believe the likelihood of this is rare, but real.
    - The solution here for security-conscious users would be to change their udev rules to allow user-access and change the dbus service to run as an unprivileged user (who has access to the USB device only).
    - This is possible today without any code change to knowledgeable users:
        a. Set up udev rules to allow unprivileged access to the device
        b. Change the user field in the `soundcraft.utils.notepad.service` file to be an unprivileged user
    - We could make this easier by helping automate this unprivileged system-bus operation mode:
        a. Add setup logic to adjust udev rules appropriately
        b. Add an optional `--run-as` parameter to the setup logic to replace the default user in `soundcraft.utils.notepad.service` when we install it

2. _`venv` or `--user` -based installs with the service set up to run as root would be bad._  Yes.  Don't do this!
    - We shouldn't ever encourage or support this, especially for `--user` installs.  In fact, there's code in the current dbus setup logic that is intended specifically to prevent this mode of operation.
    - `venv` could theoretically be fine, if the location of the virtual environment is protected and only writable by root, but this seems like a less useful mode of operation.  If you're going to go through the trouble, maybe just install globally.
    - The right solution here, I think, is to support our service to run as session-bus service when installed as `--user`.  It wouldn't take much to support this mode of operation:
        - We would need to change the location of the persistent state file so it's under ~/.local/share when running in this mode, which is something we want to do anyway to support `--no-dbus` GUI mode.
        - Like unprivileged system-bus mode, this would require udev rule changes.  It would be nice to have an automated way of helping users do this.
        - Adding a local dbus service file to launch this on the system would be easy, but we'd need to change the dbus client code to look for the service on both the session-bus and the system-bus.
        - We should document the multi-user problem, letting users know that if they want to support multiple users accessing the same hardware at the same time, they should opt for a system-bus and not session-bus installation.

About your stages...
- I think Stage 1 and 2 can be addressed by my proposals above (and below).
- Stage 3: I don't think this is something I'd be interested in doing; but I'd be happy if someone else wanted to try this, and I'd be happy to help out with the USB control scheme I reverse-engineered.
- Stage 4 and 5: :)  I won't hold my breath!

So here's what I suggest in summary:
1. In support of the `--no-dbus` GUI mode, as well as the user-owned session-bus operation, let's change the location of the statefile so it's not just hard-coded to `/var/lib/soundcraft-utils` but can also be relative to `~/.local/share/soundcraft-utils` or some other user-specified path more easily.  Maybe a config file would be useful for this.
2. In support of all unprivileged access modes, let's come up with a utility that helps users add the right udev rules to their system to provide direct access to these devices.
3. Add a flag to the system-wide setup to run as an unprivileged user (and maybe make this the default, having 'run as root' be the optional fallback)
4. Support `--user`-based installs by adding logic to set up the right session-bus dbus service information, and the client-side code to look for  both session- and system-bus objects.

How do you feel about that?

**Comment by ndim** [comment]
[This got a bit long. I fear I have forgotten a few aspects and written twice about other over the course of many hours. Sorry for that.]

> I'm not sure I agree about moving the dbus service to the session bus - This dbus service represents the real hardware which is a singleton. There's no per-user hardware. It's globally visible to the running system so the system bus fits that reality better.

There is no per PC mixer hardware either.

Let me elaborate a bit on the user interface for the audio routing state in Soundcraft Notepad mixers.

There are two parts of the audio routing state: The cables plugged into the mixer (cable state), and the audio routing to USB-audio controlled by firmware (USB audio routing state).

In a perfect world ("stage five"), the mixer hardware would have is own UI to show and change the USB audio routing state, like one LED and one pushbutton per audio routing option. Then USB audio routing state could be kept as separate from anything on a connected PC as the cable state is. This would keep the USB audio routing state right next to the cable state, completely avoiding the need to split the audio routing into cable and USB audio routing. It would just be the mixer hardware state implemented in the mixer. Now if I connect the mixer to a different PC, the complete hardware state is now available on the new PC, and no piece of software needs to be concerned about that state as the mixer hardware UI has both elements to change and show the audio routing state. However, the Notepad series of mixers is not designed with with that kind of hardware UI (either an issue of meeting a manufacturing price point, or of noting that the feature to set the audio routing could be added only after the hardware design has been finished). Also, even if future devices were changed, we still would want to support the existing devices. So, strike all that.

In a slightly less perfect world ("stage 4"), the existing mixer hardware (which has no audio routing UI whatsoever) would be combined with a mixer firmware which has both USB commands to set *and* query the audio routing state so that a proper UI for the audio routing could be implemented in PC/Mac software. That would allow a user to switch the mixer hardware between different PCs and Macs and always have the software UI to the audio routing state showing the actual hardware state. This still keeps the audio routing state at least logically connected to the cable state, even though the software UI is a bit more disconnected from the mixer hardware. Unfortunately, the Notepad series of mixers is not shipped with that kind of firmware (at least as of version 1.09).

I presume that in the Notepad series of mixers Soundcraft/Harman use a proprietary Harman DSP which might very well be a very proprietary processor architecture without any available architecture description or development tools. So I presume that patching the firmware ourselves is probably practically impossible. If we could find a direct communications channel to someone in a technical role in the Soundcraft/Harman department actually writing the firmware, we might have a chance to successfully request that feature and have it implemented. Users of the proprietary Soundcraft software for Windows/MacOS would benefit from that feature as well, after all. Normal support channels as usually established by large companies are often designed and implemented in a way well suited to evade and avoid this kind of feature request, so I would not bet that going through support could help. Soundcraft also maintain a forum, but I doubt that would help much either. Soundcraft officially only support Windows and MacOS, so unless we are prepared to argue the case that we want to plug the mixer between our multiple Windows and MacOS machines (of which I have neither), we have no case their either.

So, in the real world, the mixer hardware offers no user interface at all for the USB audio routing state, and all the user interface the user can get is software talking to the mixer via USB. However, that software cannot find out the actual hardware state, and can only set it to whatever the software UI thinks is the proper state. So the only way to keep the audio routing state set by one PC when connecting the mixer to another PC is for that other PC to not do anything regarding the audio routing state. And if we have a session dbus service instead of a system dbus service to keep the state, the only way to keep the audio routing state set by one user session when connecting the mixer to another user sesssion (on any PC) is for that other user session to not do anything regarding the audio routing state. So the user interface for every software should very clearly display the default state when its state tracker process has just come up and not set the hardware state yet being UNKNOWN.

(How do Soundcraft's Windows and MacOS software packages handler mixers with different settings being plugged into different PCs and Macs? Do they just copy the software state to the mixer on connection? Or only on user request?)

I would assume that most use cases of Notepad series mixers involve neither connecting one mixer to different PCs nor to different user sessions on the same PC, so the same session dbus would cover those cases.

Whether the other user session runs on the same PC or on a different one... is that so different?

I maintain that having a single session only dbus service should solve most cases, and is the easiest to implement. In the unlikely case we run into something which cannot be solved by a session dbus service, we can still change the session dbus service to communicate with a (to be written) system dbus service to handle different user session on the same PC, or a (to be written) across-the-LAN service to handle different user sessions on different PCs.

In any case, I think a dbus client accessing the dbus service should not be required to support more than one bus. The DBUS interface for a client should be well-defined, i.e. *either* on the system bus or on the session bus. 

Another argument against a system dbus service is that you need to deal with processes from multiple user sessions going to that same system dbus service and they can want to set the USB audio routing to a different value. Now the system dbus service needs to decide which new setting to accept, and which ones to refuse. Then you need to notify them that their state change request was refused, and those clients need to deal with that... lots of things one can simply avoid by going the session service route.

If different session dbus services have a different view on what the USB audio routing state should be, then it should not be too difficult to have the soundcraft-utils session dbus service model its device access behaviour after the well-established and well-tested pulseaudio session dbus service's model for accessing PCM audio devices. I would prefer that over inventing a new way to handle different user sessions wanting access to the same audio device which might then probably will be in conflict with established models.

> So let's talk more about your 2 security concerns, which I think are very important:
> 
>     1. _When running on the system bus as root, code errors could lead to privilege escalations._  True.  I believe the likelihood of this is rare, but real.
>        
>        * The solution here for security-conscious users would be to change their udev rules to allow user-access and change the dbus service to run as an unprivileged user (who has access to the USB device only).
>        * This is possible today without any code change to knowledgeable users:
>          a. Set up udev rules to allow unprivileged access to the device
>          b. Change the user field in the `soundcraft.utils.notepad.service` file to be an unprivileged user
>        * We could make this easier by helping automate this unprivileged system-bus operation mode:
>          a. Add setup logic to adjust udev rules appropriately
>          b. Add an optional `--run-as` parameter to the setup logic to replace the default user in `soundcraft.utils.notepad.service` when we install it

A lot of issues to consider for the system dbus service. Changing to a session dbus services avoids having to deal with all those issues.

Setup logic would need to deal with:
  * Create an unpriviledge system user (let's call that user `soundcraft-utils` for now). This is highly system specific. There are different tools to create users, and different systems have different policies as to how to call these tools in order to create a system user without special priviledges.
  * The udev rules would only need `OWNER='soundcraft-utils'` as their action. That should be common for all udev based systems, i.e. every Linux system which matters in 2020.

>     2. _`venv` or `--user` -based installs with the service set up to run as root would be bad._  Yes.  Don't do this!
>        
>        * We shouldn't ever encourage or support this, especially for `--user` installs.  In fact, there's code in the current dbus setup logic that is intended specifically to prevent this mode of operation.

Hmm. I must have overlooked that part of the code.

>        * `venv` could theoretically be fine, if the location of the virtual environment is protected and only writable by root, but this seems like a less useful mode of operation.  If you're going to go through the trouble, maybe just install globally.

Agreed.

>        * The right solution here, I think, is to support our service to run as session-bus service when installed as `--user`.  It wouldn't take much to support this mode of operation:
>          
>          * We would need to change the location of the persistent state file so it's under ~/.local/share when running in this mode, which is something we want to do anyway to support `--no-dbus` GUI mode.
>          * Like unprivileged system-bus mode, this would require udev rule changes.  It would be nice to have an automated way of helping users do this.

Not users. Admins.

>          * Adding a local dbus service file to launch this on the system would be easy, but we'd need to change the dbus client code to look for the service on both the session-bus and the system-bus.

My proposal is still to drop the system bus, which means just changing `SystemBus()` to `SessionBus()` in `dbus.py` and be done here.

>          * We should document the multi-user problem, letting users know that if they want to support multiple users accessing the same hardware at the same time, they should opt for a system-bus and not session-bus installation.

My proposal would be that if users want to support multiple user sessions configuring the USB audio routing at the same time, they should file a github issue with a feature request.

> About your stages...
> 
>     * I think Stage 1 and 2 can be addressed by my proposals above (and below).

I think Stage 1 should be addressed, and Stage 2 be left for a possible future which hopefully never comes.

>     * Stage 3: I don't think this is something I'd be interested in doing; but I'd be happy if someone else wanted to try this, and I'd be happy to help out with the USB control scheme I reverse-engineered.

I am not sure yet when kernel PCM audio device state is related to the USB audio routing state in the mixer, and should thusly be managed by the same logic. If there is such a case, the kernel could be the place to do it.

>     * Stage 4 and 5: :)  I won't hold my breath!

I will not either.

> So here's what I suggest in summary:
> 
>     1. In support of the `--no-dbus` GUI mode, as well as the user-owned session-bus operation, let's change the location of the statefile so it's not just hard-coded to `/var/lib/soundcraft-utils` but can also be relative to `~/.local/share/soundcraft-utils` or some other user-specified path more easily.  Maybe a config file would be useful for this.

Use `${XDG_RUNTIME_DIR}/soundcraft-utils/${n}` with `${n}` being the number in the dbus `/soundcraft/utils/notepad/${n}` object path? `${XDG_RUNTIME_DIR}` evaluates to `/run/user/${UID}` on both my Fedora and Raspbian systems.

Fallback locations in case `XDG_RUNTIME_DIR` is not set could be `${HOME}/.cache/soundcraft-utils/${n}` and `${HOME}/.local/lib/soundcraft-utils/${n}`. The `soundcraft-utils` directory would be created with `0700`, while the parent dirs would be created with `0755`.

The locations for a system dbus service could be very different if `XDG_RUNTIME_DIR` is not set... perhaps `/tmp/soundcraft-utils/${n}`?

>     2. In support of all unprivileged access modes, let's come up with a utility that helps users add the right udev rules to their system to provide direct access to these devices.

    #!/bin/sh
    cp contrib/udev-rules/70-soundcraft-utils.rules /etc/udev/rules.d/

Tested on Fedora 32 and Raspbian 10.4. Interestingly, the Raspbian installation uses ACLs for those permissions even though the `acl` package containing the user interface to POSIX ACLs `getfacl` and `setfacl` are not installed. So you only notice the `+` in the `ls -l` output unless you `apt-get install acl`:

    # ls -l /dev/bus/usb/001/
    [...]
    crw-rw-r--  1 root root 189, 32 Sep  2 22:10 033
    crw-rw-r--+ 1 root root 189, 33 Sep  2 22:10 034
    crw-rw-r--  1 root root 189, 34 Sep  2 22:10 035
    # 

(I use an Aten US424 four device USB 2.0 hub with four switchable PC host connections to switch the enseble of keyboard, two mice, and the mixer between the PC running Fedora and the Raspberry Pi running Raspbian, so a few switching events have happened during testing, and thus high device numbers have been reached.)

>     3. Add a flag to the system-wide setup to run as an unprivileged user (and maybe make this the default, having 'run as root' be the optional fallback)

You need also to create that unpriviledged user. The way to do that is system specific.

Having root as a fallback is dangerous. Running as root should require an explicit decision, not just happen if the unpriviledged user happens to be unavailable. *IF* you need a system dbus service, I would have `User=soundcraft-utils` in its `.service` description, having it not start if that user is unavailable.

This is one of the issues which can be avoided by switching to a session dbus service.

>     4. Support `--user`-based installs by adding logic to set up the right session-bus dbus service information, and the client-side code to look for  both session- and system-bus objects.

As I understand it, whether a dbus service is on the system or on the session bus is basically part of the DBUS API of that service.

> How do you feel about that?

I still think going for just a session dbus service ("stage 1") should be enough. Less logic, fewer things to worry about, and it can still be extended in case anyone complains about their use case being unsupported.

BTW, am I correctly assuming one important reason for the install and uninstall code residing inside the `soundcraft_dbus_service` executable is to find out the path where `soundcraft_dbus_service` is actually installed? At this time, I am still following the idea that somewhere within the egg_install logic, there must be the code which determines where to install the scripts generated for the `console_scripts` and `gui_scripts` entry points. That code should be callable in some way, so that we can put the installation of the dbus-1 and XDG files into the directories corresponding to either `/usr/local` install or `/usr` install or `$HOME` install (those three installation types appear to be the only three supported options for the future).

Having tried to extend setuptools and `setup.py` to support installation of `dbus-1` config files and XDG desktop and XDG icon files and having failed so far in a number of approaches, I have to say that distutils and setuptools appear to be written with the express purpose to be very inflexible and difficult to extend. The Automake+Autoconf m4 macro plus sh code djungle is an absolute paradise compared to that :-) Anyway, this is Python, and `setup.py` is what we have, so use `setup.py` we must.

**Comment by lack** [comment]
You said:
> I think Stage 1 should be addressed, and Stage 2 be left for a possible future which hopefully never comes.

I think I'm coming around to your way of thinking.  Moving to session bus instead of the system bus.  It does solve a lot of the permission and non-system-wide-install issues very cleanly, and the only aspect it fails to address (multiple users using the same hardware) isn't necessarily a reasonable use-case to support anyway.

(Though for the record, the current system dbus implementation actually does support this multi-user case today!  It accepts every routing change from any user, and already uses dbus signals to notify all user UIs of changes)  I'll just swallow my pride here and agree that even though it was fun to support this, it's really not that important :)

So let's focus on what needs to be done to get there:
1. Move the state file out of /var/lib and into somewhere user-specific.  I don't think `$XDG_RUNTIME_DIR` is appropriate, since at least on the systems I've looked at, this is backed by tmpfs and doesn't survive reboot.  This isn't just for the running state of the dbus, but to remember the intended state of the mixer across host reboots (since this isn't available to be queried from the mixer).  I think `${XDG_DATA_HOME}/soundcraft/state.json` may be a better place (see https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html for a description of XDG_DATA_HOME and what to do if it's not set).
    - When the session-based service starts up, we should probably import any existing config from the older /var/lib location, if it exists, but only on the first time (ie, only if the old location exists and the new location does not yet exist)
2. Change client and server to use the session bus, not the system bus
3. Change `soundcraft_dbus_service --setup` utility so that:
    1. If the system service is still registered, try to use sudo to run itself `soundcraft_dbus_service` in a mode to remove the old system dbus setup files and terminate the old system dbus service.
        - If this fails, this is not fatal to the setup, but should give specific instructions on what an admin should do manually to clean up
    2. Use sudo to create the new udev rules
        - If this fails, this is not fatal to the setup, but should give specific instructions on what an admin should do manually to set this up
    3. Sets up the session dbus service in "the right way" depending on whether this is a system-wide or `--user` install
    4. Sets up the XDG application and icon in "the right way" depending on whether this is a system-wide or `--user`
4. While we're at it, let's change all the filenames and service names over to `io.github.soundcraft-utils` for consistency.
5. Since this is a pretty major change to the architecture, I think a major version bump to 1.0.0 would be in order.

And as to why the setup code is in the soundcraft_dbus_service?  This is mainly because, as you've found out, Python's setup.py isn't designed or intended to manage anything outside of the python libraries (with the minor exception of the entrypoint scripts).  I really think if distros want to repackage this themselves (there is an AUR package for this), this is much better managed by a distro-specific packager, but I wanted something that would be easy to bootstrap using pip by itself because I don't have time to manage a bunch of distro packages.

I wonder if it would be worth considering a snap or flatpak mechanism to package this thing up, with the special considerations of course required because we need write-access to USB devices from inside of what is otherwise a pretty clean sandbox in those modes of distribution...  But that's maybe for another day's discusison.

**Comment by ndim** [comment]
> You said:
> 
> > I think Stage 1 should be addressed, and Stage 2 be left for a possible future which hopefully never comes.
> 
> I think I'm coming around to your way of thinking. Moving to session bus instead of the system bus. It does solve a lot of the permission and non-system-wide-install issues very cleanly, and the only aspect it fails to address (multiple users using the same hardware) isn't necessarily a reasonable use-case to support anyway.

I hope I have been both making sense from the technical perspective, and avoiding communicating in a too pushy or even outright rude manner. I know Germans are known for a manner of communication in the anglosaxon world and America especially which is often received bordering on rude, often from the wrong side of that border. I really hope I am avoiding that. (If not, I will appreciate if you tell me in a rudely direct manner. :-)

In any case, I am very glad that we are in agreement here.

> (Though for the record, the current system dbus implementation actually does support this multi-user case today! It accepts every routing change from any user, and already uses dbus signals to notify all user UIs of changes) I'll just swallow my pride here and agree that even though it was fun to support this, it's really not that important :)

(The multi-user case also implements an avenue for denial of service when any user can tell the system dbus service to change the USB audio routing while someone else is actually using the PCM audio devices of the mixer and expects the USB audio routing to stay a certain way. That is another problem going to a session bus service and using the well-established device access control implemented by the logic behind the `uaccess` `TAG`. :)

> So let's focus on what needs to be done to get there:
> 
>     1. Move the state file out of /var/lib and into somewhere user-specific.  I don't think `$XDG_RUNTIME_DIR` is appropriate, since at least on the systems I've looked at, this is backed by tmpfs and doesn't survive reboot.  This isn't just for the running state of the dbus, but to remember the intended state of the mixer across host reboots (since this isn't available to be queried from the mixer).  I think `${XDG_DATA_HOME}/soundcraft/state.json` may be a better place (see https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html for a description of XDG_DATA_HOME and what to do if it's not set).

I had not thought about that state needing to last across system reboots. As you cannot tell whether a mixer has been unplugged and replugged, possibly with different USB audio routing set from another PC, across a PC reboot, I would have handled that case the same as a freshly plugged in Notepad mixer: State "UNKNOWN". That would be perfectly represented by the absence of a file in `XDG_RUNTIME_DIR`. But that would then lose the information on the last desired hardware state from the point of view of the software UI, which would be bad.

So the dbus service would need to communicate to the GUI/CLI software both that the hardware state is `UNKNOWN`, but that the last set state was, say, `STEREO 7/8` and then the GUI/CLI software could offer a 'set mixer to last known state' function.

I am confusing myself a bit here right now.

>        * When the session-based service starts up, we should probably import any existing config from the older /var/lib location, if it exists, but only on the first time (ie, only if the old location exists and the new location does not yet exist)

To be honest, I would not have wasted much effort on conserving user settings when those amount to a single selection from a list of four options, especially in a piece of software with a version number starting with `0.4`. If it was more complex state, or a long established piece of software in daily use by millions of people, I would   On the other hand, this is relatively trivial logic, and it should not have consequences for other parts of the code, so I am neutral on this.

>     2. Change client and server to use the session bus, not the system bus

Ack. That is a three liner.

>     3. Change `soundcraft_dbus_service --setup` utility so that:

Could we move all that setup install/uninstall code to, say `soundcraft_setup`? That would

  * remove install/uninstall code from `dbus.py` where it does not really belong
  * concentrate the install/uninstall code in a single script which could be run when building a distro package to install the dbus and xdg files, and then not included in the distro package at all in order not to confuse distro package users about needing to run setup themselves. This requires adding support for a `--root` directory.
  * As this is still a `console_scripts` entrypoint, the script is still installed in the same location.

>        1. If the system service is still registered, try to use sudo to run itself `soundcraft_dbus_service` in a mode to remove the old system dbus setup files and terminate the old system dbus service.
>           
>           * If this fails, this is not fatal to the setup, but should give specific instructions on what an admin should do manually to clean up

Not only remove the service config file, but also quit/kill the process. Hmm. Renaming `soundcraft_dbus_service` could help with avoiding accidentally starting the new session service executable from an old system service config file as root.

Spitting out a few well written (and thus well readable) shell commands could be both documentation for the sysadmin and cut-and-pastable. Not sure whether this idea holds water, though.

>        2. Use sudo to create the new udev rules
>           
>           * If this fails, this is not fatal to the setup, but should give specific instructions on what an admin should do manually to set this up

Call me old fashioned but is "copy this file to /etc/udev/rules.d/" too complicated? Hmm. OK, you probably would also want to call `udevadm control --reload` and possibly some `udevadm` command to explicitly re-run the udev rules for the mixer hardware. If that re-running of the udev rules for the mixer hardware could be triggered from a script, that would be sweet.

Just copying a file to someplace is easy to communicate to a soundcraft-utils user and local Linux admin, but explicitly triggering the rules for just the mixers is relatively difficult to communicate would be *very* sweet to do automagically. OK, you just sold me on this setup tool.

>        3. Sets up the session dbus service in "the right way" depending on whether this is a system-wide or `--user` install

That means dropping a single file into `share/dbus-1/services/*.service`.

>        4. Sets up the XDG application and icon in "the right way" depending on whether this is a system-wide or `--user`

While we're at that, we could also fix the PNG icons to be actually installed in the proper size (we need to generate those PNG files first, though). I have a script for that, but no robust icon install method for hooking it into.

>     4. While we're at it, let's change all the filenames and service names over to `io.github.soundcraft-utils` for consistency.

Ack.

BTW, many DBUS APIs use names like `org.freedesktop.login1` containing a number for basically versioning the API (see e.g. `busctl list`).

>     5. Since this is a pretty major change to the architecture, I think a major version bump to 1.0.0 would be in order.

I would go with `0.4.1` for the current state of affairs, then implement these changes and call it `0.5.0` for the time being. Then I would wait a bit for the technical dust to settle from all these changes (there will be some), go through a few more iterations like `0.6.0` while fixing the things that are bound to come up, and finally jump to `1.0.0` in a few months when it has become clear that this architecture and DBUS API can be supported for the next few years as there have not been any significant bug reports (or feature requests about things like multi-user use cases :-) for a few months after `0.8.3`.

> And as to why the setup code is in the soundcraft_dbus_service? This is mainly because, as you've found out, Python's setup.py isn't designed or intended to manage anything outside of the python libraries (with the minor exception of the entrypoint scripts). I really think if distros want to repackage this themselves (there is an AUR package for this), this is much better managed by a distro-specific packager, but I wanted something that would be easy to bootstrap using pip by itself because I don't have time to manage a bunch of distro packages.

I would say `soundcraft-utils` should support the three install locations /usr and /usr/local (both dbus and the xdg desktop/icon stuff look in those two by default) and `$HOME/.local` via the `XDG_*` variables with hardcoded fallbacks. That helps everybody:
  * admin/user installing system wide from git source
  * user installing to `$HOME`
  * package maintainer creating distro packages
  * people using distro packages
It may even help with flatpak stuff.

However, the question is whether the extra logic can be shoehorned into `setup.py` in a way that is compatible with
  * `setup.py install`
  * `setup.py install --user`
  * `setup.py develop --user`
  * `pip install --user`
and possibly a few more ways.

If this does not fit into `setup.py`, we need a separate install/uninstall tool. (We might be able to copy what Quodlibet does inside `setup.py`.)

> I wonder if it would be worth considering a snap or flatpak mechanism to package this thing up, with the special considerations of course required because we need write-access to USB devices from inside of what is otherwise a pretty clean sandbox in those modes of distribution... But that's maybe for another day's discusison.

Flatpaks sound like something to be considered after the first dust has settled.

**Comment by ndim** [comment]
Perhaps this huge amount of changes could be made inside a branch `preparing-0.5` or `session-dbus` or `devel`, separate from `release`. I guess this will take a bit until all the parts described above are in place and fitting together well enough, and having a working `release` branch in the mean time is probably a good idea.

**Comment by lack** [comment]
I totally agree; a rewrite like this will need to be a big rewrite that basically goes in all at once.

I like calling the branch `session-dbus`.  I've just created that at the current point of the `release` branch.  Feel free to point PRs there for this effort if you're feeling like starting :)

**Comment by ndim** [comment]
> > So let's focus on what needs to be done to get there:
> > 1. Move the state file out of /var/lib and into somewhere user-specific.  I don't think `$XDG_RUNTIME_DIR` is appropriate, since at least on the systems I've looked at, this is backed by tmpfs and doesn't survive reboot.  This isn't just for the running state of the dbus, but to remember the intended state of the mixer across host reboots (since this isn't available to be queried from the mixer).  I think `${XDG_DATA_HOME}/soundcraft/state.json` may be a better place (see https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html for a description of XDG_DATA_HOME and what to do if it's not set).
> 
> I had not thought about that state needing to last across system reboots. As you cannot tell whether a mixer has been unplugged and replugged, possibly with different USB audio routing set from another PC, across a PC reboot, I would have handled that case the same as a freshly plugged in Notepad mixer: State "UNKNOWN". That would be perfectly represented by the absence of a file in `XDG_RUNTIME_DIR`. But that would then lose the information on the last desired hardware state from the point of view of the software UI, which would be bad.

I just noticed something: How do we know that the Mixer plugged in right now is the same one which has been unplugged earlier? I can see no serial numbers or other identifying information on my Mixer's `lsusb -v` output. There might be a per-device unique identifier such as the serial number available via some proprietary USB query, but that will not show up in `lsusb`, of course.

So `soundcraft-utils` is keeping the state around from the last time *some* Notepad mixer was plugged, and then just assumes the *next* Notepad mixer to be plugged in to be the same one which had been plugged in earlier, and then happily applies the former mixer's state to the new one. This works if there is only one device ever to be connected, or at least always the same model with the same external cabling and dial state.

I imagine the most common use case for a Notepad series mixers is when it is the only Notepad series mixer in the house, and thus the only Notepad series mixer any PC in that house will ever see. For that case, storing the mixer state for when the same mixer is reconnected the next time is very useful and probably an expected feature.  I am quite sure that *I* will not buy another Notepad-series mixer, as I only have a need for at most one, so this is my use case.

However, I have a feeling that the scenario of more than one mixer existing in a single house and thus being connected to the same PC if not concurrently, then at least sequentially, is quite likely to happen some time. In that case, storing the state of mixer A and then applying it to mixer B would be a mistake. Not a huge one, but a mistake. It might even be solved by documenting this behaviour. A use case for this could be someone with an 8fx for podcasting and a 12fx for their band. Or a `soundcraft-utils` maintainer having multiple devices around for testing. :-)

At this time, we can just document `soundcraft-utils` does not support more than one mixer at the same time or at any time. Not sure how far that will carry into the future, though.

Changing the state file name from `0` to something containing the device name would allow `soundcraft-utils` to store the mixer state for the use case of no two mixers of the same model ever being connected to the same PC. It still does not handle multiple mixers connected at the same time.

Not sure how to solve/manage/deal with this in the absence of both a "query USB audio routing" USB command and a per-device unique ID.

**Comment by lack** [comment]
Fixed in https://github.com/socratools/socranop/pull/4
